### ISSUE TITLE
[BUG] environment-lines are overwritten in /etc/systemd/system/kubelet.service.d/10-kubeadm.conf

### ISSUE THREAD

**Issue by rdxmb** [geerlingguy/ansible-role-kubernetes#44]
With the lines
https://github.com/geerlingguy/ansible-role-kubernetes/blob/master/tasks/kubelet-setup.yml#L16-L18
some important stuff seems to be overwritten in 
`/etc/systemd/system/kubelet.service.d/10-kubeadm.conf`

Don't know why, but I have *one* Ubuntu-Node which has no
`/etc/default/kubelet`

But the real problem is that multiple files begin with `Environment` and the first of them will be overwritten then:

correct:
```
# cat /etc/default/kubelet 
KUBELET_EXTRA_ARGS=
# cat /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
# Note: This dropin only works with kubeadm and kubelet v1.11+
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"
# This is a file that "kubeadm init" and "kubeadm join" generates at runtime, populating the KUBELET_KUBEADM_ARGS variable dynamically
EnvironmentFile=-/var/lib/kubelet/kubeadm-flags.env
# This is a file that the user can use for overrides of the kubelet args as a last resort. Preferably, the user should use
# the .NodeRegistration.KubeletExtraArgs object in the configuration files instead. KUBELET_EXTRA_ARGS should be sourced from this file.
EnvironmentFile=-/etc/default/kubelet
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBELET_KUBEADM_ARGS $KUBELET_EXTRA_ARGS
```

wrong (after changed by the lines linked above)
```
# cat /etc/default/kubelet
cat: /etc/default/kubelet: No such file or directory
# cat /etc/systemd/system/kubelet.service.d/10-kubeadm.conf 
# Note: This dropin only works with kubeadm and kubelet v1.11+
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf"
Environment="KUBELET_EXTRA_ARGS="
# This is a file that "kubeadm init" and "kubeadm join" generates at runtime, populating the KUBELET_KUBEADM_ARGS variable dynamically
EnvironmentFile=-/var/lib/kubelet/kubeadm-flags.env
# This is a file that the user can use for overrides of the kubelet args as a last resort. Preferably, the user should use
# the .NodeRegistration.KubeletExtraArgs object in the configuration files instead. KUBELET_EXTRA_ARGS should be sourced from this file.
EnvironmentFile=-/etc/default/kubelet
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBELET_KUBEADM_ARGS $KUBELET_EXTRA_ARGS
```

so the config-file `/var/lib/kubelet/config.yaml` will not be included here, which means that the dns is missing

`  Warning  MissingClusterDNS  2m17s (x260 over 62m)  kubelet, zwei  kubelet does not have ClusterDNS IP configured and cannot create Pod using "ClusterFirst" policy. Falling back to "Default" policy.`

```
# kubelet --version
Kubernetes v1.13.5
# cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
```

**Comment by rdxmb** [comment]
Line https://github.com/geerlingguy/ansible-role-kubernetes/blob/master/tasks/kubelet-setup.yml#L18 should be more explizitely, something like

`kubelet_args_regexp: '^Environment=KUBELET_EXTRA_ARGS'`

Actually I have deleted my damaged node and have little time, so I cannot create a PR nor run a test.

**Comment by rdxmb** [comment]
The problem is also, that the file `/etc/default/kubelet` is not longer installed by the ubuntu-package, so https://github.com/geerlingguy/ansible-role-kubernetes/blob/master/tasks/kubelet-setup.yml#L2-L5 does not work anymore.
https://github.com/kubernetes/release/issues/654

**Comment by rdxmb** [comment]
If somebode has a PR I can run a test against some hosts.

**Comment by nickmaccarthy** [comment]
+1 to this bug.  This is also affecting me to.  In my case the `kubeadm init` fails because we are missing `Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"` in the systemd file.  I am running ubuntu 16.04 and trying to install kubernetes stable-1.13.  What I did for now was turned the kubet-setup.yml tasks into a template that installs the kubelet service file in /etc/systemd and added a new handler to restart the systemd/kublet if it changes.  If i can get some time I try to fix this in a better way and make a PR, but perhaps a template for the systemd kubelet service might be the best option after all?

**Comment by rdxmb** [comment]
I would not overwrite the systemd-unit if it is not really necessary. Instead, I would prefer the solution to create and use `/etc/default/kubelet` with this role - at least on Debian/Ubuntu-based systems.

As it is given in https://github.com/kubernetes/release/issues/654#issue-386270494 , the motivation for not installing the file with the debian-package anymore is: `This file exists for users to customize.`

So this is what the ansible-role does here: **customizing**.

Also, this should prevent any problems caused by changing the systemd-units by the maintainers - as we just had with docker in https://github.com/docker/for-linux/issues/476

**Comment by geerlingguy** [comment]
I believe this was partially fixed in #46 / #42. However I also have open https://github.com/geerlingguy/ansible-role-kubernetes/issues/4 which is meant to use the newer config layout (this role was originally written in the 1.10 era).

Please follow / share on that issue for further improvements / changes from the env-var-based settings.
